app/models/conversion_host.rb

```diff
@@ -124,28 +208,61 @@ def resource_supports_conversion_host
     end
   end
 
-  def check_resource_credentials(fatal = false, extra_msg = nil)
-    success = send("check_resource_credentials_#{resource.ext_management_system.emstype}")
-    if !success and fatal
-      msg = "Credential not found for #{resource.name}."
-      msg += " #{extra_msg}" unless extra_msg.blank?
-      _log.error(:msg)
+  # Find the associated credentials for the current conversion host based
+  # on the type of provider. Each provider should look for a directly
+  # associated authentication first, and default to the authentication
+  # for the resource second.
+  #--
+  # TODO: Each provider subclass should define their own method for
+  # finding credentials.
+  #
+  def find_credentials(msg = nil)
+    authentication = send("find_resource_credentials_#{resource.ext_management_system.emstype}")
+
+    unless authentication
+      msg = "Credentials not found for conversion host #{name} or resource #{resource.name}"
+      msg << " #{msg}" if msg
+      _log.error(msg)
       raise MiqException::Error, msg
     end
-    success
+
+    authentication
   end
 
-  def check_resource_credentials_rhevm
-    !(resource.authentication_userid.nil? || resource.authentication_password.nil?)
+  # Find the credentials for the associated Redhat resource. If an AuthUseridPassword
+  # object is not found, but the associated userid and password are found, return
+  # a (temporary, unsaved) AuthUseridPassword object using those values.
+  #--
+  # TODO: Move this into the provider specific subclass.
+  #
+  def find_credentials_redhat
+    authentication = AuthUseridPassword.find_by(:resource => resource)
+
+    if authentication.nil? && resource.authentication_userid && resource.authentication_password
+      authentication = AuthUseridPassword.new(
```

Comment:
Why is a method called `find_` creating new authentications?  Also this won't create the record without .save!

Reply:
Right, I'm deliberately not saving it. The goal is a common object that I can then call methods on, instead of dealing with different return types depending on the provider, which is clumsy.

Reply:
...interesting could we just return the `resource.authentication_type()`

Reply:
This is still an issue, why not just return the resource authentication instead of creating a new one?

Reply:
Mainly this was done for openstack, where we want an authentication independent of the provider. But even for other providers, I felt it was more flexible to have a separate authentication.

@fdupont-redhat Am I incorrect?

Reply:
I don't mean why have a separate auth for a conversation host in general, I mean here where you're essentially just using the resources authentication just return that one like you do with openstack below instead if creating a temporary object here

Reply:
Ok, updated. I used the infra host.rb file in the ovirt provider code as a guideline.